Add camera support

anishsane opened the pull request:
It uses the mjpeg stream over the http server to access the camara feed.
The stream implementation is based on the mjpeg integration in home assistant.

anishsane commented:
This needs an update in the tasmota integration of the home assistant to work end-to-end.

emontnemery commented on hatasmota/camera.py:
```diff
@@ -0,0 +1,141 @@
+
+    def get_still_image_stream(self, websession:aiohttp.ClientSession) -> web.StreamResponse | None:
+        """Get the io stream to read the static image."""
+        still_image_url = f"http://{self._cfg.ip_address}/snapshot.jpg"
```
Is the URL always `/snapshot.jpg`, or can it be configured?

anishsane replied:
From the [Tasmota code](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_81_esp32_webcam.ino#L1110) (or [new version](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_81_esp32_webcam_task.ino#2228)), we have a choice of `/snapshot.jpg`, `/wc.jpg`, `/wc.mjpeg`. But IMO, `snapshot.jpg` seems the best description of the functionality here.

emontnemery replied:
OK, but just to clarify: Is there a difference between `/snapshot.jpg`, `/wc.jpg` and `/wc.mjpeg`?

anishsane replied:
There is no difference. All these URLs have the same callback function and return the same data.
I guess the Tasmota authors added them only for convenience.

emontnemery commented on hatasmota/camera.py:
```diff
@@ -0,0 +1,141 @@
+
+    def get_mjpeg_stream(self, websession:aiohttp.ClientSession) -> web.StreamResponse | None:
+        """Get the io stream to read the mjpeg stream."""
+        mjpeg_url = f"http://{self._cfg.ip_address}:81/cam.mjpeg"
```
Same question, is the port always 81 and the url always `cam.mjpeg` for the mjpeg stream?

anishsane replied:
From the [Tasmota code](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_81_esp32_webcam.ino#L1110) (or [new version](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_81_esp32_webcam_task.ino#189)),
The port is fixed to 81.

we have a choice of `/cam.mjpeg`, `/cam.jpg`, `/stream`. IMO, `stream` and `cam.mjpeg` are good choices among these. I chose `cam.mjpeg`, because it clarifies that the stream is an `mjpeg` stream.

emontnemery replied:
Same question as above; all the URLs just call the same function `HandleWebcamMjpeg`, does it mean there's no difference, it's just aliases for exactly the same funcitonality?

anishsane replied:
There is no difference. All these URLs have the same callback function and return the same data.
I guess the Tasmota authors added them only for convenience.

emontnemery reviewed (commented):
Looks good, just questions about configurability

emontnemery commented:
Please fix the lint errors 👍

emontnemery commented:
There are still lint errors.
I'd recommend to install and run the linters locally, let me know if you need guidance for that.

anishsane commented:
My bad.
I tested in HA directly which already had it installed.
Fixed.

anishsane commented:
I had originally added the additional constants to support resolution/brightness/contrast configuration.
I was under the assumption that camera core integration has built-in support for such config knobs.

I implemented it via home assistant blueprints instead: https://github.com/tasmota/blueprints/pull/4

emontnemery commented:
> I had originally added the additional constants to support resolution/brightness/contrast configuration. I was under the assumption that camera core integration has built-in support for such config knobs.
> 
> I implemented it via home assistant blueprints instead: [tasmota/blueprints#4](https://github.com/tasmota/blueprints/pull/4)

Hmm, I'm not sure what you mean now. Can you elaborate a bit?

emontnemery commented:
Linters are still failing..

I'll explain how to set up the linters locally then.
First time, set up a local development environment for hatasmota:
```
cd hatasmota
python3 -m venv .venv
source .venv/bin/activate
pip3 install -r requirements.txt
pip3 install -r requirements-test.txt
```

After that, activate the local development environment for hatasmota:
```
cd hatasmota
source .venv/bin/activate
```

Run the linters:
```
flake8 hatasmota
pylint hatasmota
mypy hatasmota
black hatasmota
isort hatasmota --combine-as --force-sort-within-sections --profile black
```

emontnemery commented:
The mypy error is genuine, `websession.get` returns an async context manager, not a `StreamResponse`.

I think it would make sense to modify the code like this:
```diff
diff --git a/hatasmota/camera.py b/hatasmota/camera.py
index 25fa14b..2cb4840 100644
--- a/hatasmota/camera.py
+++ b/hatasmota/camera.py
@@ -5,7 +5,7 @@ from __future__ import annotations
 from dataclasses import dataclass
 import logging
 from typing import Any
-from aiohttp import ClientSession, web
+from aiohttp import ClientResponse, ClientSession

 from .const import (
     CONF_IP,
@@ -122,12 +122,12 @@ class TasmotaCamera(TasmotaAvailability, TasmotaEntity):
         """Unsubscribe to all MQTT topics."""
         self._sub_state = await self._mqtt_client.unsubscribe(self._sub_state)

-    def get_still_image_stream(self, websession: ClientSession) -> web.StreamResponse | None:
+    async def get_still_image_stream(self, websession: ClientSession) -> ClientResponse:
         """Get the io stream to read the static image."""
         still_image_url = f"http://{self._cfg.ip_address}/snapshot.jpg"
-        return websession.get(still_image_url)
+        return await websession.get(still_image_url)

-    def get_mjpeg_stream(self, websession: ClientSession) -> web.StreamResponse | None:
+    async def get_mjpeg_stream(self, websession: ClientSession) -> ClientResponse:
         """Get the io stream to read the mjpeg stream."""
         mjpeg_url = f"http://{self._cfg.ip_address}:81/cam.mjpeg"
-        return websession.get(mjpeg_url)
+        return await websession.get(mjpeg_url)
```

You could also skip the await, and type the return values as `Awaitable[ClientResponse]`

anishsane commented:
Oh.. I had only used the pylint linter locally.
The other linters from your comment suggested changes.
Updated the PR.

About my blueprint-related comment:
My initial commit had some extra string constants added and imported. They were for configuring the camera settings like resolution. I had added them but I was not able to use them. I removed them from hatasmota and was able get that functionality using a home-assistant blueprint.

emontnemery commented on hatasmota/discovery.py:
```diff
@@ -392,6 +395,21 @@ def get_fan_entities(
     return fan_entities
 
 
+def get_camera_entities(
+    discovery_msg: dict,
+) -> list[tuple[TasmotaCameraConfig | None, DiscoveryHashType]]:
+    """Generate camera configuration."""
+    camera_entities: list[tuple[TasmotaCameraConfig | None, DiscoveryHashType]] = []
+
+    entity = None
+    discovery_hash = (discovery_msg[CONF_MAC], "cam", "cam", 0)
+    if CONF_CAM in discovery_msg and discovery_msg[CONF_CAM]:
+        entity = TasmotaCameraConfig.from_discovery_message(discovery_msg, "camera")
+    camera_entities.append((entity, discovery_hash))
+
+    return camera_entities
+
+
```
Please move this to before `get_cover_entities` to keep the functions mostly sorted alphatebetically

anishsane replied:
Done.

emontnemery commented on hatasmota/discovery.py:
```diff
@@ -486,6 +504,8 @@ def get_entities_for_platform(
         entities.extend(get_status_sensor_entities(discovery_msg))
     elif platform == "switch":
         entities.extend(get_switch_entities(discovery_msg))
+    elif platform == "camera":
+        entities.extend(get_camera_entities(discovery_msg))
```
Please move this before cover to keep the alphabetical sorting

anishsane replied:
Done.

emontnemery commented on hatasmota/discovery.py:
```diff
@@ -514,6 +534,8 @@ def get_entity(
         return TasmotaStatusSensor(config=config, mqtt_client=mqtt_client)
     if platform == "switch":
         return TasmotaRelay(config=config, mqtt_client=mqtt_client)
+    if platform == "camera":
+        return TasmotaCamera(config=config, mqtt_client=mqtt_client)
```
Please move this before cover to keep the alphabetical sorting

anishsane replied:
Done.

emontnemery commented on hatasmota/camera.py:
```diff
@@ -0,0 +1,131 @@
+class TasmotaCameraConfig(TasmotaAvailabilityConfig, TasmotaEntityConfig):
+    """Tasmota camera configuation."""
+
+    idx: int
```
Does tasmota support more than a single camera? If not, can we skip this, it seems to be hardcode to 0 below?

anishsane replied:
Removed.

emontnemery commented on hatasmota/camera.py:
```diff
@@ -0,0 +1,131 @@
+
+    idx: int
+
+    command_topic: str
```
This doesn't seem to be used

anishsane replied:
It was added earlier to support the configuring the camera parameters. But since it is not used, I have removed it.
The configuration is done easily via the blueprint mentioned earlier.

anishsane replied:
Yes. A number entity for resolution, brightness, and contrast.
We also have switch entities for flip horizontal/vertical.

Overall it looks like this with my [blueprint](https://github.com/tasmota/blueprints/pull/4):
![image](https://github.com/user-attachments/assets/ea8d4e8b-0a3e-4547-8433-ec68644a12ee)

emontnemery replied:
OK. You could add this support to hatasmota and the tasmota integration if you want, as number entities and switch entities which are discovered if there's a camera.

emontnemery commented on hatasmota/camera.py:
```diff
@@ -0,0 +1,131 @@
+    async def subscribe_topics(self) -> None:
+        """Subscribe to topics."""
+
+        def state_message_received(msg: ReceiveMessage) -> None:
```
What is this used for? The core PR doesn't seem to use it?

anishsane replied:
It was part of the original version of the core PR. It was removed during the review.
I have updated this PR to reflect that.

emontnemery reviewed (approved):
Thanks, @anishsane 👍
